File: src/main/java/io/github/alxiw/hello/service/AccountService.java
package io.github.alxiw.hello.service;

import io.github.alxiw.hello.data.AccountDaoImpl;
import io.github.alxiw.hello.model.Account;

import java.util.List;

public class AccountService {

    private static volatile AccountService instance;

    public static AccountService getInstance() {
        // Первый проверка (не блокирующая)
        if (instance == null) {
            synchronized (AccountService.class) {
                // Второй проверка (блокирующая)
                if (instance == null) {
                    instance = new AccountService();
                }
            }
        }
        return instance;
    }

    private final AccountDao accountDao;

    private AccountService() {
        this.accountDao = new AccountDaoImpl();
    }

    public Response addAccount(String uin, String name) {
        return getResponse(accountDao.addAccount(uin, name));
    }

    public Account getAccountById(int id) {
        return accountDao.getAccountById(id);
    }

    public List<Account> getAllAccounts() {
        return accountDao.getAllAccounts();
    }

    public Response updateAccount(Account account) {
        return getResponse(accountDao.updateAccount(account));
    }

    public Response deleteAccount(int id) {
        return getResponse(accountDao.deleteAccount(id));
    }

    private Response getResponse(int value) {
        if (value <= -1) {
            return Response.ERROR;
        } else if (value == 0) {
            return Response.NO_CHANGES;
        } else {
            return Response.SUCCESS;
        }
    }
}

File: src/main/java/io/github/alxiw/hello/service/account/AccountService.kt
package io.github.alxiw.hello.service.account

import io.github.alxiw.hello.data.AccountDao
import io.github.alxiw.hello.data.model.Account
import io.github.alxiw.hello.sys.Response
import io.github.alxiw.hello.sys.Utils.getDatabaseResponse
import kotlin.concurrent.Volatile

class AccountService private constructor() {

    private val accountDao: IAccountDao = AccountDao()

    fun addAccount(uin: String, name: String): Response {
        return getDatabaseResponse(accountDao.addAccount(uin, name))
    }

    fun getAccountById(id: Int): Account? {
        return accountDao.getAccountById(id)
    }

    fun getAllAccounts(): List<Account> {
        return accountDao.getAllAccounts()
    }

    fun updateAccount(account: Account): Response {
        return getDatabaseResponse(accountDao.updateAccount(account))
    }

    fun deleteAccount(id: Int): Response {
        return getDatabaseResponse(accountDao.deleteAccount(id))
    }

    companion object {

        @Volatile
        private var instance: AccountService? = null

        @JvmStatic
        fun getInstance(): AccountService {
            return instance ?: AccountService().also { instance = it }
        }
    }
}
